include/cuco/detail/hyperloglog/hyperloglog.cuh

```diff
@@ -0,0 +1,340 @@
+   * <tt>std::is_convertible<std::iterator_traits<InputIt>::value_type,
+   * T></tt> is `true`
+   *
+   * @param first Beginning of the sequence of items
+   * @param last End of the sequence of items
+   * @param stream CUDA stream this operation is executed in
+   */
+  template <class InputIt>
+  void add(InputIt first, InputIt last, cuco::cuda_stream_ref stream)
+  {
+    this->add_async(first, last, stream);
+    stream.synchronize();
+  }
+
+  /**
+   * @brief Asynchronously merges the result of `other` estimator into `*this` estimator.
+   *
+   * @tparam OtherScope Thread scope of `other` estimator
+   * @tparam OtherAllocator Allocator type of `other` estimator
+   *
+   * @param other Other estimator to be merged into `*this`
+   * @param stream CUDA stream this operation is executed in
+   */
+  template <cuda::thread_scope OtherScope, class OtherAllocator>
+  void merge_async(hyperloglog<T, Precision, OtherScope, Hash, OtherAllocator> const& other,
+                   cuco::cuda_stream_ref stream) noexcept
+  {
+    this->merge_async(other.ref(), stream);
+  }
+
+  /**
+   * @brief Merges the result of `other` estimator into `*this` estimator.
+   *
+   * @note This function synchronizes the given stream. For asynchronous execution use
+   * `merge_async`.
+   *
+   * @tparam OtherScope Thread scope of `other` estimator
+   * @tparam OtherAllocator Allocator type of `other` estimator
+   *
+   * @param other Other estimator to be merged into `*this`
+   * @param stream CUDA stream this operation is executed in
+   */
+  template <cuda::thread_scope OtherScope, class OtherAllocator>
+  void merge(hyperloglog<T, Precision, OtherScope, Hash, OtherAllocator> const& other,
+             cuco::cuda_stream_ref stream)
+  {
+    this->merge_async(other, stream);
+    stream.synchronize();
+  }
+
+  /**
+   * @brief Asynchronously merges the result of `other` estimator reference into `*this` estimator.
+   *
+   * @tparam OtherScope Thread scope of `other` estimator
+   *
+   * @param other Other estimator reference to be merged into `*this`
+   * @param stream CUDA stream this operation is executed in
+   */
+  template <cuda::thread_scope OtherScope>
+  void merge_async(ref_type<OtherScope> const& other, cuco::cuda_stream_ref stream) noexcept
+  {
+    auto constexpr block_size = 1024;
+    cuco::hyperloglog_ns::detail::merge<<<1, block_size, 0, stream>>>(other, this->ref());
+  }
+
+  /**
+   * @brief Merges the result of `other` estimator reference into `*this` estimator.
+   *
+   * @note This function synchronizes the given stream. For asynchronous execution use
+   * `merge_async`.
+   *
+   * @tparam OtherScope Thread scope of `other` estimator
+   *
+   * @param other Other estimator reference to be merged into `*this`
+   * @param stream CUDA stream this operation is executed in
+   */
+  template <cuda::thread_scope OtherScope>
+  void merge(ref_type<OtherScope> const& other, cuco::cuda_stream_ref stream)
```

Comment:
```suggestion
  template <typename Other>
  void merge(Other const& other, cuco::cuda_stream_ref stream)
  {
    static_assert(std::is_same_v<Hash, typename Other::Hash>);
    static_assert(this->precision == Other::precision);
    ...
  }
```
One overload with build time checks should be sufficient. The same for the `merge_async` variants